connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanResult.java

```diff
@@ -68,45 +71,88 @@ public PluginScanResult(
                           connectorClientConfigPolicies);
     }
 
-    public Collection<PluginDesc<SinkConnector>> sinkConnectors() {
+    /**
+     * Merge one or more {@link PluginScanResult results} into one result object
+     */
+    public PluginScanResult(List<PluginScanResult> results) {
+        this(
+                merge(results, PluginScanResult::sinkConnectors),
+                merge(results, PluginScanResult::sourceConnectors),
+                merge(results, PluginScanResult::converters),
+                merge(results, PluginScanResult::headerConverters),
+                merge(results, PluginScanResult::transformations),
+                merge(results, PluginScanResult::predicates),
+                merge(results, PluginScanResult::configProviders),
+                merge(results, PluginScanResult::restExtensions),
+                merge(results, PluginScanResult::connectorClientConfigPolicies)
+        );
+    }
+
+    private static <T, R extends Comparable<R>> SortedSet<R> merge(List<T> results, Function<T, SortedSet<R>> accessor) {
+        SortedSet<R> merged = new TreeSet<>();
+        for (T element : results) {
+            merged.addAll(accessor.apply(element));
+        }
+        return merged;
+    }
```

Comment:
Since this is private and unlikely to be used elsewhere, we don't have to make it generic (which also makes it trickier to read):
```suggestion
    private static <R extends Comparable<R>> SortedSet<R> merge(List<PluginScanResult> results, Function<PluginScanResult, SortedSet<R>> pluginsAccessor) {
        SortedSet<R> merged = new TreeSet<>();
        for (PluginScanResult element : results) {
            merged.addAll(pluginsAccessor.apply(element));
        }
        return merged;
    }
```